pkg/controller/replica.go

```diff
@@ -162,11 +165,33 @@ func (r *ReconcileReplica) triggerGarbageCollection(volumeName string) {
 func (r *ReconcileReplica) removeGarbageCollection(volumeName string) {
 	v, ok := r.cleanUpMap.LoadAndDelete(volumeName)
 	if ok {
-		cancelFunc := v.(context.CancelFunc)
-		cancelFunc()
+		entry := v.(cleanUpEntry)
+		entry.cancelFunc()
 	}
 }
 
+func (r *ReconcileReplica) retryManageReplicas(ctx context.Context, volumeName string) {
+	// retry replica management with exponential backoff until success
+	conditionFunc := func() (bool, error) {
+		// we should not run replica management while garbage collection is scheduled for the volume
+		if v, cleanUpScheduled := r.cleanUpMap.Load(volumeName); cleanUpScheduled {
+			entry := v.(cleanUpEntry)
+			select {
+			case <-entry.ctx.Done():
+				// if gargage collection cancelled, start replica management
+				break
+			case <-timeSinceEquals(entry.initTimestamp, r.timeUntilGarbageCollection):
```

Comment:
If we're waiting to start garbage collection, why not simply cancel it? (This `select` statement blocks until cancelation or start of garbage collection due to the lack of `default` case. Is this intentional?)

Reply:
This was intended as it occurred to me that we don't want to start replica management when we have garbage collection. And that is why `retryManageReplica` always wraps everything in a goroutine as we don't want to block the reconciler workers.

Reply:
> If garbage collection is pending, are you waiting for it to start in the hope that it will be canceled?

Yes, that was what I initially had in mind. Once garbage collection gets scheduled for a volume, it leaves the controller with two options: either 1) garbage collection gets executed after timer completes or 2) garbage collection gets cancelled after primary attachment gets created.

And given that we create replica attachments only under two scenarios: 1) when replica attachment gets deleted and 2) when replica attachment gets promoted (i.e. `attachment.Spec.RequestedRole != attachment.Status.Role`), and that we create at most 1 replicas per `manageReplica` call, it didn't occur to me that the controller would call `retryManageReplica` that frequently. 

But I do see your concern, and I also see another potential problem given the backoff we have added. Given that garbage collection scheduler waits 5 minutes until initiating garbage collection process, a relplicaManagement cycle that had failed numbers of time and has backoff of `> 5 minutes`, it can end up starting replica management on its next retry without realizing that garbage collection had been executed for the volume.

Reply:
I wonder if we should feed `retryManageReplica` with a `context.WithCancelFunc`, which gets cancelled when garbage collection is initiated. Then, once garbage collection kicks in, 1) the `ExponentialBackoffWithContext` would stop retrying and 2) `manageReplica` will not be able to create replicas with cancelled context.